src/library/scala/util/Random.scala

```diff
@@ -70,6 +70,16 @@ class Random(val self: java.util.Random) extends AnyRef with Serializable {
    */
   def nextLong(): Long = self.nextLong()
 
+  def nextLong(n: Long): Long = {
+    if (n <= 0) throw new java.lang.IllegalArgumentException("bound must be positive")
+    if (n <= Int.MaxValue) nextInt(n.toInt).toLong
+    else {
+      val RAND_MAX = Long.MaxValue     // 2^63-1
+      val x = nextLong() >>> 1         // [0, 2^63), strip the sign bit
+      if (x < (RAND_MAX - (RAND_MAX % n))) x % n else nextLong(n)
```

Comment:
Here is a purely functional version with tail-recursion:
```scala
else {
  val RandMax = Long.MaxValue
  val bound = RandMax - (RandMax % n)
  @inline @tailrec def genValid(): Long = {
    val x = nextLong() >>> 1
    if (x < bound) x else genValid()
  }
  genValid() % n
}
```
The comparisons are not inefficient. Modulo operations are. Even on the JVM, they are relatively very slow compared to other arithmetic operations. On JS they are *terribly* slow, especially when the left-hand-side is > 2^53, which is true for the first `%` and virtually always true for the second `%`.